internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+func (m *TLSManager) genECP256Keypair(notBefore time.Time, notAfter time.Time) (*TLSKeyPair, error) {
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	// Random serial for now.
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		NotBefore:    notBefore,
+		NotAfter:     notAfter,
+		Subject: pkix.Name{
+			Organization: []string{"Canonical Ltd."},
+		},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+	// DER encoded bytes
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, err
+	}
+	keypair := &TLSKeyPair{privateKey: privateKey, Certificate: certificate}
+	keypair.addFingerprint()
+	return keypair, nil
+}
+
+// persistTLSKeyPair persists the TLS keypair in a single PEM file. The file name
+// is the certificate fingerprint string, with a .pem suffix. See the
+// addFingerprint method for information on how the fingerprint is generated.
+func (m *TLSManager) persistTLSKeyPair(keypair *TLSKeyPair) (err error) {
+	// If the TLS directory does not yet exist, create it.
+	_, err = os.Stat(m.tlsDir)
+	if os.IsNotExist(err) {
+		// Create the leaf directory node with 0700 permissions.
+		err = os.Mkdir(m.tlsDir, 0700)
+	}
+	if err != nil {
+		return fmt.Errorf("cannot create directory leaf of path %v: %w", m.tlsDir, err)
+	}
+
+	pemPath := filepath.Join(m.tlsDir, fmt.Sprintf("%v.pem", keypair.Fingerprint()))
+	pemFile, err := os.OpenFile(pemPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+
+	pemData, err := keypair.toPEMBytes()
+	if err != nil {
+		return err
+	}
+	if _, err = pemFile.Write(pemData); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TLSKeyPair represents a TLS key pair (see tls.LoadX509KeyPair), which
+// includes both the x509 certificate and the private key (unexported). This
+// type can directly be converted to a tls.Certificate for configuring a TLS
+// listener, or exported as a PEM encoded x509 certificate for easy transport
+// over the daemon API.
+type TLSKeyPair struct {
+	privateKey  any
+	Certificate *x509.Certificate
+	fingerprint string
+}
+
+// Compare compares two keypairs and returns an int as used by sorting
+// functions.
+func (t TLSKeyPair) Compare(other *TLSKeyPair) int {
```

Comment:
This function compares keypairs in a very specific way, and I don't think we should consider that _the_ canonical way to compare keypairs. In fact, I'm not sure it's correct in our use case to compare by start time then by end time. What about end time first then start time? Either way there are situations where some clients will lose access even though they still hold a "valid" certificate.

In short, I think this function should have a name that suggest that it is specifically for ordering keypairs in order of creation (I suppose that's the intent) and not a universally valid comparison.

Reply:
This is a pattern in Go used for sorting. We agreed with CTO we will order by start time for now.

Reply:
PR reworked. N/A for current PR. Resolving.